spec/controllers/devise/saml_sessions_controller_spec.rb

```diff
@@ -17,25 +17,79 @@ def require_no_authentication
 
 require_relative '../../../app/controllers/devise/saml_sessions_controller'
 
-
 describe Devise::SamlSessionsController, type: :controller do
   let(:saml_config) { Devise.saml_config }
+  let(:idp_providers_adapter) { spy("Stub IDPSettings Adaptor") }
+
+  before do
+    allow(idp_providers_adapter).to receive(:settings).and_return({
+      assertion_consumer_service_url: "acs_url",
+      assertion_consumer_service_binding: "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST",
+      name_identifier_format: "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress",
+      issuer: "sp_issuer",
+      idp_entity_id: "http://www.example.com",
+      authn_context: "",
+      idp_slo_target_url: "http://idp_slo_url",
+      idp_sso_target_url: "http://idp_sso_url",
+      idp_cert: "idp_cert"
+    })
+  end
 
   describe '#new' do
-    it 'redirects to the SAML Auth Request endpoint' do
-      get :new
-      expect(response).to redirect_to(%r(\Ahttp://localhost:8009/saml/auth\?SAMLRequest=))
+    let(:saml_response) { File.read(File.join(File.dirname(__FILE__), '../../support', 'response_encrypted_nameid.xml.base64')) }
+
+    context "when using the default saml config" do
+      it "redirects to the IdP SSO target url" do
+        get :new, "SAMLResponse" => saml_response
+        expect(response).to redirect_to(%r(\Ahttp://localhost:8009/saml/auth\?SAMLRequest=))
+      end
+    end
+
+    context "with a specified idp" do
```

Comment:
In this new context, `new` redirects to the SSO target url of the IdP that corresponds to the `Issuer` in the parsed SAML when `Devise.idp_settings_adapter` is set.